STM32: Add SPI RTIO support

gautierg-st opened the pull request:
Add SPI RTIO support for STM32. This requires interrupts (no DMA yet)

gautierg-st committed:
drivers: spi: stm32: prepare for rtio integration

Move some code into a new function to prepare for RTIO integration.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>
Signed-off-by: Etienne Carriere <etienne.carriere@st.com>

decsny commented:
is there some benchmark showing any value add over the default implementation?

gautierg-st committed:
tests: drivers: spi: loopback: add stm32 spi rtio test case

Add a test case for STM32 SPI RTIO.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

erwango reviewed (changes requested):
First round of review

mathieuchopstm commented on drivers/spi/Kconfig.stm32:
```diff
@@ -49,4 +50,22 @@ config SPI_STM32_BUSY_FLAG_TIMEOUT
+	int "Number of available submission queue entries"
+	default 8 # Sensible default that covers most common spi transactions
+	help
+	  When RTIO is use with SPI each driver holds a context with which blocking
```
And `with` should be dropped, as the full sentence is otherwise not very sensical:

"When RTIO is use with SPI each driver holds a context _with_ which blocking API calls use to perform SPI transactions."

mathieuchopstm reviewed (commented):
Seems OK

gautierg-st committed:
tests: drivers: spi: loopback: boards: adjust scaling for stm32 boards

Adjust transfer duration scaling config for STM32 boards

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

etienne-lms committed:
drivers: spi: stm32: don't include zephyr/arch/cache.h

Remove inclusion of zephyr_arch/cache.h header file from STM32 SPI
driver. This header file is included by zephyr/cache.h if applicable
(e.g. CONFIG_ARCH_CACHE is enabled) and should not be used when
CONFIG_EXTERNAL_CACHE is enabled.

Signed-off-by: Etienne Carriere <etienne.carriere@st.com>

gautierg-st commented on drivers/spi/Kconfig.stm32:
```diff
@@ -49,4 +50,22 @@ config SPI_STM32_BUSY_FLAG_TIMEOUT
 endif # SPI_STM32_ERRATA_BUSY
 
+if SPI_RTIO
+config SPI_STM32_RTIO_SQ_SIZE
```
Done

gautierg-st commented:
spi_loopback test is passing on all series, RTIO or not (except STM32H7, but it is not related to this PR).
@avisconti I have retested on a sensortile_box_pro and it works on my end.

avisconti commented:
> > spi_loopback test is passing on all series, RTIO or not (except STM32H7, but it is not related to this PR). @avisconti I have retested on a sensortile_box_pro and it works on my end.
> 
> OK Guillaume, I'll retest then. Maybe I did something wrong. But tomorrow

@gautierg-st 
Guillaume, I moved to 54c1d6e1b, then I did

```
west build -b sensortile_box_pro samples/sensor/stream_drdy/
west flash
```

But on my end it is not working. Did you do something different?

GeorgeCGV commented on drivers/spi/spi_ll_stm32.c:
```diff
@@ -487,12 +494,207 @@ static void spi_stm32_cs_control(const struct device *dev, bool on)
+		spi_stm32_cs_control(dev, false);
+
+		if (spi_rtio_complete(rtio_ctx, status)) {
+			err = spi_stm32_iodev_prepare_start(dev);
+			if (err == 0) {
```
Minor: `err` could be removed - `if (spi_stm32_iodev_prepare_start(dev) == 0)`.

GeorgeCGV commented on drivers/spi/spi_ll_stm32.c:
```diff
@@ -487,12 +494,207 @@ static void spi_stm32_cs_control(const struct device *dev, bool on)
+		}
+	}
+}
+#endif
```
```suggestion
#endif /* CONFIG_SPI_RTIO */
```

GeorgeCGV commented on drivers/spi/spi_ll_stm32.c:
```diff
@@ -487,12 +494,207 @@ static void spi_stm32_cs_control(const struct device *dev, bool on)
+	if (spi_stm32_iodev_complete(dev, status)) {
+		return;
+	}
+#endif
```
```suggestion
#endif /* CONFIG_SPI_RTIO */
```

GeorgeCGV commented on drivers/spi/spi_ll_stm32.c:
```diff
@@ -559,6 +761,17 @@ static void spi_stm32_isr(const struct device *dev)
+	    !LL_SPI_IsEnabledIT_TXE(spi)) {
+		return;
+	}
+#endif
```
```suggestion
#endif /* CONFIG_SPI_RTIO && !DT_HAS_COMPAT_STATUS_OKAY(st_stm32h7_spi) */
```

GeorgeCGV commented on drivers/spi/spi_ll_stm32.c:
```diff
@@ -630,6 +844,7 @@ static int spi_stm32_configure(const struct device *dev,
 		}
 		return 0;
 	}
+#endif
```
```suggestion
#endif /* CONFIG_SPI_RTIO */
```

GeorgeCGV commented on drivers/spi/spi_ll_stm32.c:
```diff
@@ -923,6 +1139,11 @@ static int transceive(const struct device *dev,
+#ifdef CONFIG_SPI_RTIO
+	ret = spi_rtio_transceive(data->rtio_ctx, config, tx_bufs, rx_bufs);
+	goto end;
+#endif
```
```suggestion
#endif /* CONFIG_SPI_RTIO */
```

gautierg-st commented:
Rebased and comments taken into account, but the issue on sensortile_box_pro is still present.

avisconti commented:
> Rebased and comments taken into account, but the issue on sensortile_box_pro is still present.

Aaaargh! :(

avisconti reviewed (dismissed):
I think it should pass a test on sensortile_box_pro, where sensors are attached in SPI.

etienne-lms commented on drivers/spi/spi_ll_stm32.c:
```diff
@@ -545,12 +552,152 @@ static void spi_stm32_msg_start(const struct device *dev, bool is_rx_empty)
+	struct spi_rtio *rtio_ctx = data->rtio_ctx;
+	bool ret = false;
+
+	if (!status && rtio_ctx->txn_curr->sqe.flags & RTIO_SQE_TRANSACTION) {
```
```suggestion
	if (status == 0 && (rtio_ctx->txn_curr->sqe.flags & RTIO_SQE_TRANSACTION= != 0) {

etienne-lms commented on drivers/spi/spi_ll_stm32.h:
```diff
@@ -183,6 +186,17 @@ static inline void ll_func_disable_int_errors(SPI_TypeDef *spi)
+static inline bool ll_func_are_int_disabled(SPI_TypeDef *spi)
+{
+#if DT_HAS_COMPAT_STATUS_OKAY(st_stm32h7_spi)
+	return (spi->IER == 0U);
```
```suggestion
	return spi->IER == 0U;
```

etienne-lms commented on drivers/spi/spi_ll_stm32.h:
```diff
@@ -183,6 +186,17 @@ static inline void ll_func_disable_int_errors(SPI_TypeDef *spi)
+#if DT_HAS_COMPAT_STATUS_OKAY(st_stm32h7_spi)
+	return (spi->IER == 0U);
+#else
+	return ((LL_SPI_IsEnabledIT_ERR(spi) ||
+		 LL_SPI_IsEnabledIT_RXNE(spi) ||
+		 LL_SPI_IsEnabledIT_TXE(spi)) == 0);
```
I would rather suggest a form easier to read:
```suggestion
	return !LL_SPI_IsEnabledIT_ERR(spi) &&
	       !LL_SPI_IsEnabledIT_RXNE(spi) &&
	       !LL_SPI_IsEnabledIT_TXE(spi);

etienne-lms committed:
drivers: spi: stm32: remove useless LOG_LEVEL

Remove useless LOG_LEVEL in STM32 SPI RTIO driver.

Signed-off-by: Etienne Carriere <etienne.carriere@st.com>

etienne-lms committed:
drivers: spi: stm32: replace an #if with an if() in transceive()

Replace a #ifndef directive with a if(!IS_ENABLED()) instrcution
in transceive() function. This change makes later integration of RTIO
support in this function smoother, polluting a bit less this function
with #if based directives.

Signed-off-by: Etienne Carriere <etienne.carriere@st.com>

etienne-lms committed:
drivers: spi stm32: remove bits2byte()

Remove bits2byte() helper function than was not always used. Replace it
with a division by BITS_PER_BYTE that is explicit enough.

Signed-off-by: Etienne Carriere <etienne.carriere@st.com>

etienne-lms committed:
drivers: spi: stm32: explicit boolean test

Use explicit boolean tests in STM32 SPI driver.

Signed-off-by: Etienne Carriere <etienne.carriere@st.com>

etienne-lms committed:
drivers: spi: stm32: correct indentation

Correct some indentation issues, a few useless line escapes,
double space characters or parentheses pair in STM32 SPI driver.

Signed-off-by: Etienne Carriere <etienne.carriere@st.com>

etienne-lms committed:
drivers: spi: stm32: remove useless inline keywords

Remove useless inline keywords in STM32 SPI driver.

Signed-off-by: Etienne Carriere <etienne.carriere@st.com>

etienne-lms committed:
drivers: spi: stm32: fix missing braces in conditioned instruction

Fix a missing braces pair around a conditioned instruction in
STM32 SPI driver. Fix that by aggregating the 2 if() instructions
into a single ANDed one.

Signed-off-by: Etienne Carriere <etienne.carriere@st.com>

etienne-lms committed:
drivers: spi: stm32: reorder header inclusion

Reorder inclusions of header files to clarify and simplify later changes.
By the way, remove #ifdef CONFIG_SPI_STM32_DMA condition to include
DMA header files as its not required.

Keep zephyr/log.h with use of LOG_MODULE_REGISTER() first since included
local spi_context.h depends on log resources to be defined.

Signed-off-by: Etienne Carriere <etienne.carriere@st.com>

etienne-lms commented:
Hi @gautierg-st, I took over your series. Tests are fine on my side on in our HW bench. The main blocking issue in the previous series regarding stream tests was a call that `spi_stm32_complete()` should not call to `spi_stm32_iodev_complete()` when there are no more sequences to work on (`spi_rtio::txn_head == NULL`).

@avisconti, I've successfully tested on sensor_box_pro.

I've appended a series of cleanup changes on the SPI driver. A few issues had to be addressed but made more changes than I initially planed. If one think it's too much and these should be dropped, tell me, I'll keep only the really useful ones.

gautierg-st reviewed (commented):
One small comment on something I did, otherwise LGTM (but I can't approve since it's originally my PR).
And congratulations again for finding what was causing problem for the stream !

etienne-lms commented on drivers/spi/spi_ll_stm32.c:
```diff
@@ -985,6 +1131,11 @@ static int transceive(const struct device *dev,
+#ifdef CONFIG_SPI_RTIO
+	ret = spi_rtio_transceive(data->rtio_ctx, config, tx_bufs, rx_bufs);
+	goto end;
+#endif /* CONFIG_SPI_RTIO */
```
Ok, I'll update.

gautierg-st committed:
drivers: spi: stm32: add rtio support

Add SPI RTIO support for STM32. SPI RTIO required interrupts.
DMA is not supported yet.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>
Signed-off-by: Etienne Carriere <etienne.carriere@st.com>

etienne-lms commented:
All good in our HW bench.

avisconti reviewed (approved):
As far as I can tell, this PR looks good to me. Thanks for your effort!
